parsl/tests/test_providers/test_local_provider.py

```diff
@@ -0,0 +1,202 @@
+    assert status.exit_code == 0
+    assert status.stdout == 'magic'
+    assert status.stderr == ''
+
+    status = _run(p, '/bin/echo -n magic 1>&2')
+    assert status.state == JobState.COMPLETED
+    assert status.exit_code == 0
+    assert status.stdout == ''
+    assert status.stderr == 'magic'
+
+
+def test_local_channel():
+    with tempfile.TemporaryDirectory() as script_dir:
+        script_dir = tempfile.mkdtemp()
+        p = LocalProvider(channel=LocalChannel(), launcher=SingleNodeLauncher(debug=False))
+        p.script_dir = script_dir
+        _run_tests(p)
+
+
+SSHD_CONFIG = """
+Port {port}
+ListenAddress 127.0.0.1
+HostKey {hostkey}
+AuthorizedKeysFile {connpubkey}
+AuthenticationMethods publickey
+StrictModes no
+Subsystem sftp {sftp_path}
+"""
+
+
+# It would probably be better, when more formalized site testing comes into existence, to
+# use a site-testing provided server/configuration instead of the current scheme
+def test_ssh_channel():
+    with tempfile.TemporaryDirectory() as config_dir:
+        sshd_thread, priv_key, server_port = _start_sshd(config_dir)
+        try:
+            with tempfile.TemporaryDirectory() as remote_script_dir:
+                # run this once so we add host keys and whatnot
+                p = subprocess.run(['ssh', '-v', '-i', priv_key, '-oStrictHostKeyChecking=no',
+                                    '-oUserKnownHostsFile={}/known.hosts'.format(config_dir),
+                                    '-p', str(server_port), '127.0.0.1'], check=False, stdout=subprocess.PIPE,
+                                   stderr=subprocess.PIPE, input='')
+                if p.returncode != 0:
+                    raise Exception('SSH failed: {}, {}, {}'.format(p.returncode, p.stdout, p.stderr))
+                script_dir = tempfile.mkdtemp()
+                p = LocalProvider(channel=SSHChannel('127.0.0.1', port=server_port,
+                                                     script_dir=remote_script_dir,
+                                                     host_keys_file='{}/known.hosts'.format(config_dir),
+                                                     key_filename=priv_key),
+                                  launcher=SingleNodeLauncher(debug=False))
+                p.script_dir = script_dir
+                _run_tests(p)
+        finally:
+            _stop_sshd(sshd_thread)
+
+
+def _stop_sshd(sshd_thread):
+    sshd_thread.stop()
+
+
+class SSHDThread(threading.Thread):
+    def __init__(self, config_file):
+        threading.Thread.__init__(self, daemon=True)
+        self.config_file = config_file
+        self.stop_flag = False
+        self.error = None
+
+    def run(self):
+        try:
+            # sshd needs to be run with an absolute path, hence the call to which()
+            p = subprocess.Popen([shutil.which('sshd'), '-D', '-f', self.config_file],
+                                 stdout=subprocess.PIPE, stderr=subprocess.PIPE)
```

Comment:
I think this use of PIPE without being read until the end will lead to blocks on pipe writes. That problem exists elsewhere in parsl - https://github.com/Parsl/parsl/issues/1628

I'm unsure if that will make sshd hang, but it wouldn't surprise me.

Reply:
This is a test and -D is "don't detach" not "debug".

Reply:
@benclifford I'm guessing this one's resolved, can you confirm please.

Reply:
I guess so. My main concern was if it made a lot of output, but as @hategan says -D is detach, not verbose debug output as I initially misread it.